pkg/operator/etcdcertsigner/etcdcertsignercontroller.go

```diff
@@ -178,24 +181,69 @@ func (c *EtcdCertSignerController) syncAllMasters(recorder events.Recorder) erro
 	return utilerrors.NewAggregate(errs)
 }
 
+// validateCertificateSAN makes sure that a given certificate is valid
+// at least for the SANs defined in the configuration.
+func (c *EtcdCertSignerController) validateCertificateData(certName string, ipAddresses []string) error {
+	secret, err := c.secretLister.Secrets(operatorclient.TargetNamespace).Get(certName)
+	if err != nil {
+		return fmt.Errorf("could not get secrets from %s: %w", certName, err)
+	}
+	if secret.Data == nil {
+		return fmt.Errorf("secret data not found in %s", certName)
+	}
+
+	return verifyCertificateSAN(secret.Data, certName, ipAddresses)
+}
+
+func verifyCertificateSAN(certData map[string][]byte, certName string, ipAddresses []string) error {
+	certificates, err := cert.ParseCertsPEM(certData["tls.crt"])
+	if err != nil {
+		return fmt.Errorf("could not parse TLS cert for %s: %w", certName, err)
+	}
+	// Using the first certificate to verify the IP addresses
+	certificate := certificates[0]
+	for _, ipAddress := range ipAddresses {
+		if err := certificate.VerifyHostname(ipAddress); err != nil {
+			return fmt.Errorf("SAN for the certificate %s does not include %s: %w", certName, ipAddress, err)
+		}
+	}
+	return nil
+}
+
 func (c *EtcdCertSignerController) createSecretForNode(node *corev1.Node, recorder events.Recorder) error {
 	etcdPeerClientCertName := tlshelpers.GetPeerClientSecretNameForNode(node.Name)
 	etcdServingCertName := tlshelpers.GetServingSecretNameForNode(node.Name)
 	metricsServingCertName := tlshelpers.GetServingMetricsSecretNameForNode(node.Name)
 
-	var err error
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdPeerClientCertName)
-	peerClientCertOk := err == nil
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(etcdServingCertName)
-	servingCertOk := err == nil
-	_, err = c.secretLister.Secrets(operatorclient.TargetNamespace).Get(metricsServingCertName)
-	metricsCertOk := err == nil
+	nodeInternalIPs, err := dnshelpers.GetInternalIPAddressesForNodeName(node)
+	if err != nil {
+		return err
+	}
+
+	errors := []error{}
+	for _, certName := range []string{etcdPeerClientCertName, etcdServingCertName, metricsServingCertName} {
+		if err := c.validateCertificateData(certName, nodeInternalIPs); err != nil {
+			klog.Errorf("validateCertificateSAN: %v", err)
+			errors = append(errors, err)
+			break
+		}
+	}
 
 	// if we have all the certs we want, do nothing.
-	if peerClientCertOk && servingCertOk && metricsCertOk {
+	if len(errors) == 0 {
 		return nil
 	}
 
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:    "EtcdCertSignerControllerDegraded",
```

Comment:
The general pattern is to return error to sync resulting in status change or we will have multiple updates on a single sync.